pkg/shortnames/shortnames.go

```diff
@@ -386,6 +384,55 @@ func Resolve(ctx *types.SystemContext, name string) (*Resolved, error) {
 	return resolved, nil
 }
 
+// canPromptWithTTY indicates if stdout or stdin is not a TTY.
+func canPromptWithTTY() bool {
+	return term.IsTerminal(int(os.Stdout.Fd())) && term.IsTerminal(int(os.Stdin.Fd()))
+}
+
+// Prompt opens a prompt in the terminal to let the user choose *one* of the
+// specified items.
+func Prompt(mode types.ShortNameMode, original string, items []string) (string, error) {
+	switch len(items) {
+	case 0:
+		return "", fmt.Errorf("cannot prompt without items")
+	case 1:
+		// NOTE: return the original will make sure that information on
+		// how the (possibly short) name is resolved afterwards. For
+		// instance, `podman-remote pull short` will then print if
+		// `short` resolved to an alias.
+		//
+		// Returning `items[0]` would return a FQN and hence hide the
+		// fact that it's an alias when actually pulling down the
+		// image.
```

Comment:
Thanks for the large comment, I would have missed this.

This feels unintuitive.
- Basically, the question is “what is the semantics of the return value?” “It’s one of the fully-qualified options in `items` that is safe to send to the remote server” is a nice and simple story; “it can be a fully-qualified option or a short name that needs to be processed $somehow” is unclear.
- The caller already has both `original` and `items`, so it can make whatever policy decisions it wants, based on any of `len(items) == 1`, `original \in items`, `original == Prompt()`, or other conditions.
  - (I wouldn’t even mind if `Prompt` just failed on `len(items) <= 1`, with the rationale that it’s a prompting utility, not a policy mechanism.)
- Compare how `Resolve` does return a `resolved` with full candidates, not some single-candidate special case.
- It seems to me that the caller already must have at least a _suspicion_ that the input is a short name or an alias, to even trigger a `PullCandidates` operation and to decide to invoke the `Prompt` UI. Whether the input is an alias or a short name is not a question `Prompt` can answer; OTOH the return value of `PullCandidates` should clearly indicate this already. (Maybe `PullCandidates` could need to return a `shortnames.rationale`-like value, but again I don’t think it’s too likely `Prompt` can help with reconstructing that if it is not provided by the server.)

There’s a probably a good reason for this, so I’m curious to see the caller PR which needs this semantics.

Reply:
The _main_ reason to have this logic in Prompt() is to avoid leaking any short-name mode logic outside of `pkg/shortnames`.  The logic of dealing with the short-name mode could be moved into Podman but that would scatter it a bit too much for my taste.

Maybe I am overly cautious?

Reply:
OK, so this is not just a prompting utility.

Ignore the above for now, and I’ll re-visit after understanding the full picture.

Reply:
https://github.com/containers/podman/pull/11867 is putting the pieces together

Reply:
Reading the Podman piece now, I don’t think this works quite correctly now: it _always_ tries to prompt (but perhaps fails in enforcing mode) if there are multiple options, but that’s not what happens locally with `ShortNameDisabled`, or non-interactive `ShortNamePermissive`.

So this should be closer to including the `switch mode` code from `Resolve`, but at a first glance the requirements look different enough that I’m not sure we can actually share that switch.

---

The `ShortName{Permissive,Disabled}` case does, though, convince me that the remote client might need to submit an unresolved short name to the server, at least in that case. (I’m not sure that the client should submit a short name in the alias case, but it might turn out to be the simplest approach.)

(I haven’t thought through what the `Resolved.Description` logging should like in the remote case; that might be a relevant design constraint; for now I’m just pointing out the need to handle `ShortName{Permissive,Disabled}`.)

Reply:
> The main reason to have this logic in Prompt() is to avoid leaking any short-name mode logic outside of pkg/shortnames.

(FWIW if the logic ends up more complex, as it probably will, it totally makes sense to have the client-side the logic here; even if it had to be duplicated with no code sharing, better to have it duplicated in a single file than across repositories.)